modules/ingest-ecs/src/main/java/org/elasticsearch/ingest/ecs/EcsNamespaceProcessor.java

```diff
@@ -0,0 +1,275 @@
+    private static final String CLOUD_PREFIX = "cloud";
+    private static final String HOST_PREFIX = "host";
+
+    private static final String ATTRIBUTES_KEY = "attributes";
+    private static final String RESOURCE_KEY = "resource";
+    private static final String SCOPE_KEY = "scope";
+    private static final String BODY_KEY = "body";
+    private static final String TEXT_KEY = "text";
+    private static final String STRUCTURED_KEY = "structured";
+
+    EcsNamespaceProcessor(String tag, String description) {
+        super(tag, description);
+    }
+
+    @Override
+    public String getType() {
+        return TYPE;
+    }
+
+    @Override
+    public IngestDocument execute(IngestDocument document) {
+        Map<String, Object> source = document.getSource();
+
+        boolean isOTel = isOTelDocument(source);
+        if (isOTel) {
+            return document;
+        }
+
+        // non-OTel document
+
+        Map<String, Object> newAttributes = new HashMap<>();
+        // The keep keys indicate the fields that should be kept at the top level later on when applying the namespacing.
+        // However, at this point we need to move their original values (if they exist) to the one of the new attributes namespaces, except
+        // for the @timestamp field. The assumption is that at this point the document is not OTel compliant, so even if a valid top
+        // level field is found, we assume that it does not bear the OTel semantics.
+        for (String keepKey : KEEP_KEYS) {
+            if (keepKey.equals("@timestamp")) {
+                continue;
+            }
+            Object value = source.remove(keepKey);
+            if (value != null) {
+                newAttributes.put(keepKey, value);
+            }
+        }
+
+        Map<String, Object> newResource = new HashMap<>();
+        Map<String, Object> newResourceAttributes = new HashMap<>();
+        newResource.put(ATTRIBUTES_KEY, newResourceAttributes);
+
+        source.put(ATTRIBUTES_KEY, newAttributes);
+        source.put(RESOURCE_KEY, newResource);
+
+        renameSpecialKeys(document);
+
+        // Iterate through all top level keys and move them to the appropriate namespace
+        final var sourceItr = source.entrySet().iterator();
+        while (sourceItr.hasNext()) {
+            final var entry = sourceItr.next();
+            final var key = entry.getKey();
+            final var value = entry.getValue();
+            if (KEEP_KEYS.contains(key)) {
+                continue;
+            }
+            if (shouldMoveToResourceAttributes(key)) {
+                newResourceAttributes.put(key, value);
+            } else {
+                newAttributes.put(key, value);
+            }
+            sourceItr.remove();
+        }
+
+        // Flatten attributes
+        source.replace(ATTRIBUTES_KEY, Maps.flatten(newAttributes, false, false));
+        newResource.replace(ATTRIBUTES_KEY, Maps.flatten(newResourceAttributes, false, false));
```

Comment:
`replace` seems like an interesting for these two lines -- why not just `put(...)`? Is there a circumstance in which those keys will not actually be present?

Reply:
No, with the current state of the code these keys should always be mapped. I'll change to `put`